src/url.rs

```diff
@@ -190,6 +190,37 @@ impl PyUrl {
         }
         cls.call1((url,))
     }
+
+    #[pyo3(signature=(path, *args, no_trailing_slash=false))]
+    pub fn join(&self, path: &str, args: &Bound<'_, PyTuple>, no_trailing_slash: bool) -> PyResult<Self> {
+        let join_path = |base: &Url, partial_path: String| -> Result<Url, PyErr> {
+            base.join(&partial_path)
+                .map_err(|err| PyValueError::new_err(err.to_string()))
+        };
+
+        let add_trailing_slash = |path_str: &mut String| {
+            if !path_str.trim_end().ends_with('/') {
+                path_str.push('/');
+            }
+        };
+
+        let num_of_args = args.len();
+        let mut path_str = path.to_string();
+        if num_of_args != 0 || !no_trailing_slash {
+            add_trailing_slash(&mut path_str);
+        }
+
+        let mut new_url = join_path(&self.lib_url, path_str)?;
+
+        for i in 0..num_of_args {
+            let mut arg_str = args.get_item(i)?.to_string();
+            if i != num_of_args - 1 || !no_trailing_slash {
+                add_trailing_slash(&mut arg_str);
+            }
+            new_url = join_path(&new_url, arg_str)?;
+        }
```

Comment:
I think rather than this implementation we can use [`Url::path_segments_mut`](https://docs.rs/url/latest/url/struct.Url.html#method.path_segments_mut) to push each segment in turn.

I think adding a trailing slash is then as simple as pushing `""` as a final step.

Reply:
I agree with you, but wouldn't this approach negate the additional features of `Url::join`, such as joining queries and fragments, and supporting path traversal?
```rust
let mut a = Url::parse("http://example.com/").unwrap();
println!("{}", a.join("a/b/../c").unwrap());

a.path_segments_mut().map_err(|_| "").unwrap().push("a/b/../c");
println!("{}", a);
```
```bash
http://example.com/a/c
http://example.com/a%2Fb%2F..%2Fc
```

I think what you are suggesting could be implemented as a separate function, such as `add_path` or something similar.

Reply:
Fair point, I had overlooked those possibilities 👍